examples/sparse_regression.py

```diff
@@ -0,0 +1,202 @@
+import argparse
+import time
+
+import numpy as onp
+
+import jax
+from jax import vmap
+from jax.config import config as jax_config
+import jax.numpy as np
+import jax.random as random
+
+import numpyro.distributions as dist
+from numpyro.handlers import sample
+from numpyro.hmc_util import initialize_model
+from numpyro.mcmc import mcmc
+
+
+"""
+We demonstrate how to do (fully Bayesian) sparse linear regression using the
+approach described in [1]. This approach is particularly suitable for situations
+with many feature dimensions (large P) but not too many datapoints (small N).
+In particular we consider a quadratic regressor of the form:
+
+f(X) = constant + theta_i X_i + theta_ij X_i X_j + observation noise
+
+References
+[1] The Kernel Interaction Trick: Fast Bayesian Discovery of Pairwise
+    Interactions in High Dimensions.
+    Raj Agrawal, Jonathan H. Huggins, Brian Trippe, Tamara Broderick
+    https://arxiv.org/abs/1905.06501
+"""
+
+
+def dot(X, Z):
+    return np.dot(X, Z[..., None])[..., 0]
+
+
+# The kernel that corresponds to our quadratic regressor.
+def kernel(X, Z, eta1, eta2, c):
+    eta1sq = np.power(eta1, 2.0)
+    eta2sq = np.power(eta2, 2.0)
+    k1 = 0.5 * eta2sq * np.power(1.0 + dot(X, Z), 2.0)
+    k2 = 0.5 * eta2sq * dot(np.power(X, 2.0), np.power(Z, 2.0))
+    k3 = (eta1sq - eta2sq) * dot(X, Z)
+    k4 = np.power(c, 2.0) - 0.5 * eta2sq
+    return k1 + k2 + k3 + k4
+
+
+# Most of the model code is concerned with constructing the sparsity inducing prior.
+def model(X, Y, hypers):
+    S, P, N = hypers['expected_sparsity'], X.shape[1], X.shape[0]
+
+    sigma = sample("sigma", dist.HalfNormal(hypers['alpha3']))
+    phi = sigma * (S / np.sqrt(N)) / (P - S)
+    eta1 = sample("eta1", dist.HalfCauchy(phi))
+
+    msq_inv = sample("msq_inv", dist.Gamma(hypers['alpha1'], hypers['beta1']))
+    msq = 1.0 / msq_inv
+    xisq_inv = sample("xisq_inv", dist.Gamma(hypers['alpha2'], hypers['beta2']))
+
+    eta2 = np.power(eta1, 2.0) * msq_inv / np.sqrt(xisq_inv)
+
+    lam = sample("lambda", dist.HalfCauchy(np.ones(P)))
+    kappa = np.sqrt(msq) * lam / np.sqrt(msq + np.power(eta1 * lam, 2.0))
+
+    kX = kappa * X
+
+    # sample observation noise
+    prec_obs = sample("prec_obs", dist.Gamma(3.0, 1.0))
+
+    # compute kernel
+    k = kernel(kX, kX, eta1, eta2, hypers['c']) + np.eye(N) / prec_obs
+    assert k.shape == (N, N)
+
+    # sample Y according to the standard gaussian process formula
+    sample("Y", dist.MultivariateNormal(loc=np.zeros(X.shape[0]), covariance_matrix=k),
+           obs=Y)
+
+
+# Compute the mean and variance of coefficient theta_i (where i = dimension) for a
+# MCMC sample (eta1, xi, ...). Compare to theorem 5.1 in reference [1].
+def compute_mean_variance(X, Y, dimension, msq_inv, lam, eta1, xisq_inv, c, prec_obs):
+    P, N = X.shape[1], X.shape[0]
+
+    probe = np.zeros((2, P))
+    probe = jax.ops.index_update(probe, jax.ops.index[:, dimension], np.array([0.5, -0.5]))
```

Comment:
this requires jax 0.1.38

Reply:
If I understand correctly, then the probe value at `dimension` should be `[1., -1.]` (in other words, `e_i = 1`). Because the kernel is not linear, I don't see the equivalence of your derivation and Theorem 5.1. Could you pls elaborate this point?

Reply:
i think it's fine because the quadratic piece cancels out